azure/scope/machine.go

```diff
@@ -220,47 +221,109 @@ func (m *MachineScope) InboundNatSpecs(portsInUse map[int32]struct{}) []azure.Re
+			spec.PublicLBName = m.OutboundLBName(m.Role())
+			spec.PublicLBAddressPoolName = m.OutboundPoolName(m.OutboundLBName(m.Role()))
+			if m.IsAPIServerPrivate() {
+				spec.InternalLBName = m.APIServerLBName()
+				spec.InternalLBAddressPoolName = m.APIServerLBPoolName(m.APIServerLBName())
+			} else {
+				spec.PublicLBNATRuleName = m.Name()
+				spec.PublicLBAddressPoolName = m.APIServerLBPoolName(m.APIServerLBName())
+			}
+		}
 
-	if m.Role() == infrav1.ControlPlane {
-		spec.PublicLBName = m.OutboundLBName(m.Role())
-		spec.PublicLBAddressPoolName = m.OutboundPoolName(m.OutboundLBName(m.Role()))
-		if m.IsAPIServerPrivate() {
-			spec.InternalLBName = m.APIServerLBName()
-			spec.InternalLBAddressPoolName = m.APIServerLBPoolName(m.APIServerLBName())
-		} else {
-			spec.PublicLBNATRuleName = m.Name()
-			spec.PublicLBAddressPoolName = m.APIServerLBPoolName(m.APIServerLBName())
+		// If NAT gateway is not enabled and node has no public IP, then the NIC needs to reference the LB to get outbound traffic.
+		if m.Role() == infrav1.Node && !m.Subnet().IsNatGatewayEnabled() && !m.AzureMachine.Spec.AllocatePublicIP {
+			spec.PublicLBName = m.OutboundLBName(m.Role())
+			spec.PublicLBAddressPoolName = m.OutboundPoolName(m.OutboundLBName(m.Role()))
 		}
-	}
 
-	// If NAT gateway is not enabled and node has no public IP, then the NIC needs to reference the LB to get outbound traffic.
-	if m.Role() == infrav1.Node && !m.Subnet().IsNatGatewayEnabled() && !m.AzureMachine.Spec.AllocatePublicIP {
-		spec.PublicLBName = m.OutboundLBName(m.Role())
-		spec.PublicLBAddressPoolName = m.OutboundPoolName(m.OutboundLBName(m.Role()))
-	}
+		if m.cache != nil {
+			spec.SKU = &m.cache.VMSKU
+		}
 
-	if m.Role() == infrav1.Node && m.AzureMachine.Spec.AllocatePublicIP {
-		spec.PublicIPName = azure.GenerateNodePublicIPName(m.Name())
+		if m.Role() == infrav1.Node && m.AzureMachine.Spec.AllocatePublicIP {
+			spec.PublicIPName = azure.GenerateNodePublicIPName(m.Name())
+		}
+		return append(nicSpecs, spec)
 	}
 
-	if m.cache != nil {
-		spec.SKU = &m.cache.VMSKU
-	}
+	for i, n := range m.AzureMachine.Spec.NetworkInterfaces {
+		if n.ID != "" {
+			continue
+		}
+		spec := &networkinterfaces.NICSpec{
+			Name:                  azure.GenerateNICName(m.Name()) + "-" + strconv.Itoa(i),
+			ResourceGroup:         m.ResourceGroup(),
+			Location:              m.Location(),
+			SubscriptionID:        m.SubscriptionID(),
+			MachineName:           m.Name(),
+			VNetName:              m.Vnet().Name,
+			VNetResourceGroup:     m.Vnet().ResourceGroup,
+			AcceleratedNetworking: n.AcceleratedNetworking,
+			IPv6Enabled:           m.IsIPv6Enabled(),
+			EnableIPForwarding:    m.AzureMachine.Spec.EnableIPForwarding,
+			IPConfigs:             []networkinterfaces.IPConfig{},
+		}
 
-	return []azure.ResourceSpecGetter{spec}
+		spec.SubnetName = n.SubnetName
+
+		// Check for control plane interface setup on interface 0
+		if m.Role() == infrav1.ControlPlane && i == 0 {
```

Comment:
Do we have a guarantee that the first `NetworkInterface` in the `m.AzureMachine.Spec.NetworkInterfaces` slice will always be eth0?

Reply:
Yeah I would definitely prefer a more reliable indication. Maybe we could make a helper function in the scope for getting eth0? That way it's less error prone and more readable.

Reply:
Can you clarify what you are proposing? Utilizing the interface index appears to be the only current reliable way of ensuring your intended interface order in the OS.

Reply:
Oh re-reading your comment I think I understand. The intent is to improve the code readability. Makes sense to me to have a helper function.